bucketlistapi/resources.py

```diff
@@ -0,0 +1,323 @@
+    '''
+    The resource helps registers a new user with the post method.
+    [POST]
+
+    params: username and password.
+    '''
+
+    @use_args(user_reg_login_field)
+    def post(self, args):
+        # call methods to save user data and return success or otherwise
+        self.username = args.get('username')
+        self.password = args.get('password')
+        if self.check_user_exists():
+            abort(400, message='username already exists')
+        if (self.check_name() and
+                self.check_password()):
+            user = self.create_new_user()
+            return user.as_dict(), 201 if self.save_user(user) else abort(
+                400, message='bad request')
+        else:
+            abort(400, message='username/password minimum length is 8')
+
+    def check_name(self):
+        return self.username.isalpha() and (len(self.username.strip()) > 7)
+
+    def check_password(self):
+        return len(self.password.strip()) > 7
+
+    def create_new_user(self):
+        # create a user an instance of the User Model
+        user = User(self.username.title().strip(), self.password)
+        user.hash_password()
+        return user
+
+    def check_user_exists(self):
+        return User.query.filter_by(
+            username=self.username.strip().title()).first()
+
+    @staticmethod
+    # save a user in the database using the save() utilitity
+    def save_user(user):
+        return save(user)
+
+
+class UserLoginAPI(Resource):
+    '''
+    Logs in a user to a session
+    '''
+
+    @use_args(user_reg_login_field)
+    def post(self, args):
+        # Return token for valid user
+        self.username = args.get('username')
+        self.password = args.get('password')
+        user = User.query.filter_by(
+            username=self.username.strip().title()).first()
+        if user and user.verify_password(self.password):
+            token = user.generate_auth_token()
+            return {'token': token.decode('ascii')}
+        else:
+            abort(404, message='Invalid username/password')
+
+
+class BucketListAPI(Resource):
+    '''
+    The class for the bucket list resource
+    POST : To create a new Bucket List
+    GET : To retrieve the list of all bucketlists.
+    GET : Retrieve a single bucket list
+    PUT: To update a single bucket list.
+    DELETE: To remove a bucket list
+
+    params:
+    [POST] bucketlist 'name', location json body
+    [PUT] buckelistname or done location json body
+    [GET] limit, name search(q), query string
+    '''
+    decorators = [auth.login_required]
+
+    def __init__(self):
+        self.created_by = int(g.user.id)
+
+    @use_args(name_field)
+    def post(self, args, bucketlist_id=None):
+        # This handles the creation of a new bucketlist
+        if bucketlist_id:
+            abort(404, message="method not supported for the URL")
+        self.name = args.get('name')  # self.parser.parse_args().get('name')
+        if not self.check_name():
```

Comment:
This is fine but we can make the control flow more readable by saving the bucketlist after checking
1.  If the bucket name is valid
2. If bucket name does not exist.
And then catching the non successful paths afterwards